src/python/netvlad/netvlad_extractor.py

```diff
@@ -0,0 +1,217 @@
+
+        # Whitening weights.
+        w = mat["net"].layers[33].weights[0]  # Shape: 1 x 1 x IN x OUT
+        b = mat["net"].layers[33].weights[1]  # Shape: OUT
+        # Prepare for PyTorch - make sure it is float32 and has right shape
+        w = torch.tensor(w).float().squeeze().permute([1, 0])  # OUT x IN
+        b = torch.tensor(b.squeeze()).float()  # Shape: OUT
+        # Update layer weights.
+        self.whiten.weight = nn.Parameter(w)
+        self.whiten.bias = nn.Parameter(b)
+
+        # Preprocessing parameters.
+        self.preprocess = {
+            "mean": mat["net"].meta.normalization.averageImage[0, 0],
+            "std": np.array([1, 1, 1], dtype=np.float32),
+        }
+
+    def forward(self, image):
+        assert image.shape[0] == 3
+        image = torch.clamp(image * 255, 0.0, 255.0)  # Input should be 0-255.
+        print("Image dimensions", image.size())
+
+        mean_tensor = image.new_tensor(self.preprocess["mean"]).view(1, -1, 1, 1)
+        std_tensor = image.new_tensor(self.preprocess["std"]).view(1, -1, 1, 1)
+
+        # Standardize the input image the same way the training dataset was standardized
+        image = image - mean_tensor
+        image = image / std_tensor
+
+        # Feature extraction.
+        # Image shape: (1, 3, 244, 244)
+        # Returns a (1, 512, 14, 14) feature volume.
+        descriptors = self.backbone(image)
+        print("Descriptors", descriptors.shape)
+        # Reshape the feature volume -> (1, 512, 14*14)
+        batch, channels, width, height = descriptors.size()
+        descriptors = descriptors.view(batch, channels, width * height)
+
+        # NetVLAD layer.
+        # Pre-normalization.
+        descriptors = nn.functional.normalize(descriptors, dim=1)
+        descriptor = self.vladLayer(descriptors)
+
+        # Whiten
+        descriptor = self.whiten(descriptor)
+        descriptor = nn.functional.normalize(descriptor, dim=1)
+        return descriptor
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--data_dir",
+        type=Path,
+        help="Path to the directory with images.",
+        required=True,
+    )
+    parser.add_argument(
+        "--output_file", type=Path, help="Output file name, .txt", required=True
+    )
+    parser.add_argument(
+        "--netvlad_weights_file",
+        type=Path,
+        help="NetVlad weights file.",
+        default="data/Pitts30K_struct.mat",
+    )
+    args = parser.parse_args()
+
+    # Get cpu or gpu device for training.
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+    print(f"Using {device} device")
+    model = NetVLAD(args.netvlad_weights_file).to(device)
+
+    netVladDescriptors = []
+    # ToTensor() transforms a numpy array to a correct tensor for network to work with.
+    dataset = ImageDataset(img_dir=args.data_dir, transform=ToTensor())
+    for image_idx in range(len(dataset)):
```

Comment:
You don't seem to need the image index, why not use a value-based for loop instead?
```python
for image in dataset:
  pass
```

Reply:
I remembered why it is like this.
I wanted to keep count of the processed images.
Will add the print for this until I haven't integrated that cool progress bar library.